### PR TITLE
feat: Add mocked NpAuth, NpWebApi responses and basic HTTP functionality

### PR DESCRIPTION
**Pull request by mercury501**
Made GetAuthorizationCode from NpAuth return a dummy code (could be exposed in configs to be used for custom authentication like a secret)
Added basic mocked responses to sceNpWebApi for friendList and blockList

Added basic http functionality to sceHttp, for now, every call is redirected to localhost.

Open to suggestions on configs, considerations and enhancements in general.

Cheers!

### CODE REVIEW

File: src/core/libraries/network/http.cpp [hunk 1]
```diff
@@ -430,48 +569,78 @@ int PS4_SYSV_ABI sceHttpParseStatusLine(const char* statusLine, u64 lineLen, int
...
+
+        return read_len;
+    }
+
     return ORBIS_OK;
 }
 
 int PS4_SYSV_ABI sceHttpRedirectCacheFlush() {
-    LOG_ERROR(Lib_Http, "(STUBBED) called");
+    LOG_ERROR(Lib_Http, "(STUBBED) called, returning zero to {}", __builtin_return_address(0));
```
Comment (StevenMiller123) [inline src/core/libraries/network/http.cpp:659]: Last I checked, built-in return address doesn't work because of the wrapper we have for HLEs.

File: src/core/libraries/network/http.cpp [hunk 2]
```diff
@@ -430,48 +569,78 @@ int PS4_SYSV_ABI sceHttpParseStatusLine(const char* statusLine, u64 lineLen, int
...
 
 int PS4_SYSV_ABI sceHttpsEnableOptionPrivate() {
-    LOG_ERROR(Lib_Http, "(STUBBED) called");
+    LOG_ERROR(Lib_Http, "(STUBBED) called, returning zero to {}", __builtin_return_address(0));
     return ORBIS_OK;
 }
 
-int PS4_SYSV_ABI sceHttpSendRequest(int reqId, const void* postData, u64 size) {
-    LOG_ERROR(Lib_Http, "(STUBBED) called reqId = {} size = {}", reqId, size);
+int PS4_SYSV_ABI sceHttpSendRequest(int req_id, const void* post_data, u64 size) {
```
Comment (StevenMiller123) [inline src/core/libraries/network/http.cpp:693]: If possible, it would be nice to properly return errors when the internet config option is disabled.

File: src/core/libraries/network/http.h [hunk 3]
```diff
@@ -26,6 +32,280 @@ struct OrbisHttpUriElement {
...
+    }
+
+    u32 ReadData(char* dest, u32 size) {
+
+        if (result_body == nullptr || dest == nullptr || size == 0 || result_body_size == 0) {
+
+            return 0;
+        }
+
+        size_t start_index = static_cast<size_t>(current_result_read_chunk_index) * size;
```
Comment (StevenMiller123) [inline src/core/libraries/network/http.h]: Instead of using types like size_t, we generally prefer using our types like u64.
This is rather loosely followed by maintainers though, so just consider this a minor nitpick.

File: src/core/libraries/network/http.h [hunk 4]
```diff
@@ -26,6 +32,280 @@ struct OrbisHttpUriElement {
...
+    bool IsSent() const {
+        return is_sent;
+    }
+
+    bool IsCompleted() {
+        return status_code != -1;
+    }
+
+    void DebugPrint() const {
+        LOG_CRITICAL(Lib_Http, "===== HTTP Request Debug Info =====");
```
Comment (StevenMiller123) [inline src/core/libraries/network/http.h]: Probably should change this to debug or trace logs if you plan to keep it.
I assume this was just for testing purposes though?

File: src/core/libraries/np/np_score.cpp [hunk 5]
```diff
@@ -233,8 +238,16 @@ int PS4_SYSV_ABI sceNpScoreRecordScoreAsync() {
     return ORBIS_OK;
 }
 
-int PS4_SYSV_ABI sceNpScoreSanitizeComment() {
+int PS4_SYSV_ABI sceNpScoreSanitizeComment(s32 request_handle, char* input_string,
+                                           char* output_buffer, s32 options) {
     LOG_ERROR(Lib_NpScore, "(STUBBED) called");
+
+    if (output_buffer == nullptr || input_string == nullptr) {
+        return ORBIS_OK;
```
Comment (StevenMiller123) [inline src/core/libraries/np/np_score.cpp]: Are you sure this case returns ORBIS_OK?
These types of errors usually cause some sort of invalid argument return, which could probably be tested on real hardware too.
(I haven't decompiled this library yet though)

File: src/core/libraries/np/np_web_api.cpp [hunk 6]
```diff
@@ -168,8 +221,22 @@ s32 PS4_SYSV_ABI sceNpWebApiGetHttpResponseHeaderValueLength() {
     return ORBIS_OK;
 }
 
-s32 PS4_SYSV_ABI sceNpWebApiGetHttpStatusCode() {
-    LOG_ERROR(Lib_NpWebApi, "(STUBBED) called");
+s32 PS4_SYSV_ABI sceNpWebApiGetHttpStatusCode(s64 request_id, s32* out_status_code) {
+    LOG_ERROR(Lib_NpWebApi, "(STUBBED) called, request_id: '{}'", request_id);
+
+    if (out_status_code == nullptr) {
```
Comment (StevenMiller123) [inline src/core/libraries/np/np_web_api.cpp:228]: Are you certain this case returns ORBIS_OK?

File: src/core/libraries/np/np_web_api.cpp [hunk 7]
```diff
@@ -219,8 +286,29 @@ s32 PS4_SYSV_ABI sceNpWebApiIntRegisterServicePushEventCallbackA() {
     return ORBIS_OK;
 }
 
-s32 PS4_SYSV_ABI sceNpWebApiReadData() {
-    LOG_ERROR(Lib_NpWebApi, "(STUBBED) called");
+s32 PS4_SYSV_ABI sceNpWebApiReadData(s64 request_id, char* data, u64 size) {
+    LOG_ERROR(Lib_NpWebApi, "(STUBBED) called, request_id: '{}'", request_id);
+
+    if (data == nullptr || size == 0) {
```
Comment (StevenMiller123) [inline src/core/libraries/np/np_web_api.cpp:293]: Again, correct me if you've actually tested and confirmed this (or decompiled and found this), but are you sure this case returns ORBIS_OK?

**Review by StevenMiller123 — CHANGES_REQUESTED** [review]
Couple minor nitpicks, but overall this looks really promising.
I haven't tested yet, but I should be able to test later today.

File: src/core/libraries/network/http.cpp (hunk 1 above)
Comment (mercury501) [reply src/core/libraries/network/http.cpp:659]: `[Lib.Http] <Error> http.cpp:619 sceHttpsEnableOption: (STUBBED) called, returning zero to 0x801e7dae3`
I don't have a log for that specific function, but it looks like it's working, I can remove them if you think they're superfluous

Comment (StevenMiller123) [reply src/core/libraries/network/http.cpp:659]: If it works then it is probably fine.

File: src/core/libraries/network/http.cpp (hunk 2 above)
Comment (mercury501) [reply src/core/libraries/network/http.cpp:693]: Ok so something like ORBIS_HTTP_ERROR_NETWORK ? I can add it to the sceHttpInit, since it should always be called before any other function

File: src/core/libraries/network/http.h (hunk 4 above)
Comment (mercury501) [reply src/core/libraries/network/http.h]: If you think it's ok I'd make it debug level, yeah it's for testing but could be useful for future work

File: src/core/libraries/np/np_score.cpp (hunk 5 above)
Comment (mercury501) [reply src/core/libraries/np/np_score.cpp]: Not really sure, that was to emulate the "stubbed" behavior if arguments were invalid
the changes to the Np libs I made to pass bloodborne's online functionality checks (this is the profanity check for character name)

File: src/core/libraries/np/np_web_api.cpp (hunk 6 above)
Comment (mercury501) [reply src/core/libraries/np/np_web_api.cpp:228]: Tested with bloodborne, accepts the two mocked calls (blockList and friendList)

it should work like sceHttp, so get status code accepts it in out_status_code if ORBIS_OK, since bb is in polling on it after sending the async request, ORBIS_HTTP_ERROR_EAGAIN signals that the call isn't done yet.

File: src/core/libraries/np/np_web_api.cpp (hunk 7 above)
Comment (mercury501) [reply src/core/libraries/np/np_web_api.cpp:293]: Same logic as the other one, with invalid input I just used the "stubbed" behavior

File: src/core/libraries/network/http.cpp (hunk 2 above)
Comment (StevenMiller123) [reply src/core/libraries/network/http.cpp:693]: It shouldn't be in sceHttpInit, the network errors only occur in functions like this that would actually send data

File: src/core/libraries/network/http.h (hunk 4 above)
Comment (StevenMiller123) [reply src/core/libraries/network/http.h]: I think future work does benefit from this sort of logging, but having it spam logs will reduce performance and make logs less readable.
Having it as a debug log allows us to enable it when we're testing it, while these logs remain disabled by default.

File: src/core/libraries/np/np_score.cpp (hunk 5 above)
Comment (StevenMiller123) [reply src/core/libraries/np/np_score.cpp]: Ideally we should return proper errors when we encounter them.
I'd recommend decompiling the actual library, or writing hardware tests to get an idea of how the library behaves in these edge cases.

File: src/core/libraries/network/http.cpp (hunk 2 above)
Comment (mercury501) [reply src/core/libraries/network/http.cpp:693]: ok, should I add it to the functions I've implemented or every function in sceHttp?

File: src/core/libraries/np/np_score.cpp (hunk 5 above)
Comment (mercury501) [reply src/core/libraries/np/np_score.cpp]: ok I'll take a look at 
sceNpScoreSanitizeComment
sceNpWebApiGetHttpStatusCode
sceNpWebApiReadData
and add the errors I find

File: src/core/libraries/network/http.cpp (hunk 2 above)
Comment (StevenMiller123) [reply src/core/libraries/network/http.cpp:693]: Just functions you've implemented that would error in this case.
From my own testing, I know functions like sceHttpAddRequestHeader don't throw these errors.

**Comment by mercury501** [comment]
Added correct invalid argument errors, let me know if there's anything else that needs changing!

**Comment by StevenMiller123** [comment]
Seems you're still failing clang formatting.
A lot of your empty lines have extra tabs in them, which seems to cause most of the fails.

Some IDEs have built-in clang formatting support, like VSCode. Alternatively, you can make a git patch using the output of the format pipeline and apply that.

**Comment by mercury501** [comment]
I think I've formatted them now, from what I can see if sceHttpInit fails it defaults to ORBIS_HTTP_ERROR_BEFORE_INIT, and every other function checks for init as first thing. I'm inclined on handling g_isConnectedToNetwork as ORBIS_HTTP_ERROR_BEFORE_INIT
(bloodborne just checks for a non zero result to error out)

**Comment by StevenMiller123** [comment]
Bloodborne really isn't a great test for these things, though I understand that's what you were aiming to improve in this PR.
If you're not set up to test libSceHttp with internet disabled, I can probably make a homebrew to test that at some point.

**Comment by mercury501** [comment]
by the way, all the changes I've tested with bb and my custom server I'm developing for the game

**Comment by mercury501** [comment]
> Bloodborne really isn't a great test for these things, though I understand that's what you were aiming to improve in this PR. If you're not set up to test libSceHttp with internet disabled, I can probably make a homebrew to test that at some point.

tbh I haven't ever developed a ps4 homebrew yet, if you could do that it'd be great

**Comment by mercury501** [comment]
Should I leave it as a draft then?

**Comment by StevenMiller123** [comment]
For now that's probably ideal, at least until we can get some testing info to see how well this works in more games.
I'm currently wrestling with PS4 issues, once I'm done with that I'll be able to properly test this with my library and online-only titles.

**Comment by StevenMiller123** [comment]
No regressions with these changes so far. Here's the logs from the games I've tested (with some filters to help focus the logs on Np/Http code). 

[CUSA09072.log](https://github.com/user-attachments/files/24267405/CUSA09072.log)
[CUSA07713.log](https://github.com/user-attachments/files/24267409/CUSA07713.log)
[CUSA03234.log](https://github.com/user-attachments/files/24267456/CUSA03234.log)
[CUSA44116.log](https://github.com/user-attachments/files/24267460/CUSA44116.log)
[CUSA06548.log](https://github.com/user-attachments/files/24267478/CUSA06548.log)
[CUSA05209.log](https://github.com/user-attachments/files/24267486/CUSA05209.log)
[CUSA14210.log](https://github.com/user-attachments/files/24267502/CUSA14210.log)
[CUSA00093.log](https://github.com/user-attachments/files/24267519/CUSA00093.log)
[CUSA07022.log](https://github.com/user-attachments/files/24267520/CUSA07022.log)
[CUSA27946.log](https://github.com/user-attachments/files/24267523/CUSA27946.log)
[CUSA29813.log](https://github.com/user-attachments/files/24267532/CUSA29813.log)
[CUSA14395.log](https://github.com/user-attachments/files/24267576/CUSA14395.log)
[CUSA35049.log](https://github.com/user-attachments/files/24267587/CUSA35049.log)

File: src/core/libraries/np/np_web_api.cpp [hunk 8]
```diff
@@ -123,8 +134,42 @@ s32 PS4_SYSV_ABI sceNpWebApiCreateMultipartRequest() {
     return ORBIS_OK;
 }
 
-s32 PS4_SYSV_ABI sceNpWebApiCreateRequest() {
+s32 PS4_SYSV_ABI sceNpWebApiCreateRequest(s32 title_user_ctx_id, const char* p_api_group,
+                                          const char* p_path, s32 method,
```
Comment (georgemoralis) [inline src/core/libraries/np/np_web_api.cpp:138]: method can be replaced with 

enum OrbisNpWebApiHttpMethod : s32 {
    ORBIS_NP_WEBAPI_HTTP_METHOD_GET = 0,
    ORBIS_NP_WEBAPI_HTTP_METHOD_POST = 1,
    ORBIS_NP_WEBAPI_HTTP_METHOD_PUT = 2,
    ORBIS_NP_WEBAPI_HTTP_METHOD_DELETE = 3,
};

File: src/core/libraries/np/np_web_api.h [hunk 9]
```diff
@@ -11,6 +13,15 @@ class SymbolsResolver;
 
 namespace Libraries::Np::NpWebApi {
 
+struct __attribute__((packed)) SceNpWebApiContentParameter {
+    uint64_t content_length;
```
Comment (georgemoralis) [inline src/core/libraries/np/np_web_api.h:17]: u64

File: src/core/libraries/np/np_web_api.h [hunk 10]
```diff
@@ -11,6 +13,15 @@ class SymbolsResolver;
...
+    const char* p_content_type;
+    uint8_t reserved[16];
```
Comment (georgemoralis) [inline src/core/libraries/np/np_web_api.h:19]: u8

File: src/core/libraries/np/np_score.cpp [hunk 11]
```diff
@@ -233,8 +239,17 @@ int PS4_SYSV_ABI sceNpScoreRecordScoreAsync() {
     return ORBIS_OK;
 }
 
-int PS4_SYSV_ABI sceNpScoreSanitizeComment() {
+int PS4_SYSV_ABI sceNpScoreSanitizeComment(s32 request_handle, char* input_string,
+                                           char* output_buffer, s32 options) {
```
Comment (georgemoralis) [inline src/core/libraries/np/np_score.cpp]: void* option

File: src/core/libraries/np/np_web_api.cpp (hunk 7 above)
Comment (StevenMiller123) [reply src/core/libraries/np/np_web_api.cpp:293]: This case should return `ORBIS_NP_WEB_API_INVALID_ARGUMENT`, I've spent some time decompiling this library to check.

File: src/core/libraries/np/np_web_api.cpp [hunk 12]
```diff
@@ -123,8 +134,42 @@ s32 PS4_SYSV_ABI sceNpWebApiCreateMultipartRequest() {
...
+        type = REQ_FRIEND_LIST;
+    }
+
+    if (type == REQ_INVALID) {
+
+        LOG_ERROR(Lib_NpWebApi, "No mock for request path: '{}'", p_path);
+        return ORBIS_OK;
+    }
+
+    g_requests.emplace(request_id, type);
```
Comment (StevenMiller123) [inline src/core/libraries/np/np_web_api.cpp:171]: Mocking behavior should probably be moved to `sceNpWebApiSendRequest`, `sceNpWebApiSendRequest2`, `sceNpWebApiSendMultipartRequest`, and `sceNpWebApiSendMultipartRequest2`, to better align with how the actual library functions.

These four functions return `ORBIS_NP_WEB_API_ERROR_NOT_SIGNED_IN` when PSN is disconnected.

**Comment by georgemoralis** [comment]
closing it as it already implemented in different ways
